server/controllers/caseNotesController.ts

```diff
@@ -20,63 +29,332 @@ export default class CaseNotesController {
+      const caseNote = {
+        type,
+        subType,
+        text,
+        date,
+        hours,
+        minutes,
+      }
+      const errors = this.validate(type, subType, text, date, hours, minutes)
+      if (!errors.length) {
+        try {
+          await this.caseNotesService.addCaseNote(res.locals.user.token, prisonerNumber, caseNote)
+        } catch (error) {
+          if (error.status === 400) {
+            errors.push({ text: error.message, href: '' })
+          } else throw error
+        }
+      }
+
+      if (errors.length) {
+        req.flash('caseNote', caseNote as never)
+        req.flash('errors', errors as never[])
+        req.flash('addCaseNoteRefererUrl', refererUrl)
+        return res.redirect(`/prisoner/${prisonerNumber}/add-case-note`)
+      }
+
+      if (type === 'REPORTS' && subType === 'REP_IEP') {
+        return res.redirect(
+          `${config.serviceUrls.digitalPrison}/prisoner/${prisonerNumber}/add-case-note/record-incentive-level`,
+        )
+      }
+
+      req.flash('flashMessage', { text: 'Case note added', type: FlashMessageType.success } as never)
+      return res.redirect(refererUrl || `/prisoner/${prisonerNumber}/case-notes`)
     }
+  }
 
-    // Get total count of case notes ignoring filters
-    const caseNotesUsage = await this.prisonApiClientBuilder(clientToken).getCaseNotesUsage(req.params.prisonerNumber)
-    const hasCaseNotes = Array.isArray(caseNotesUsage) && caseNotesUsage.length
+  /**
+   * Map CaseNoteType array into structures suitable for using in type > subtype dropdowns.
+   *
+   * @param caseNoteTypes
+   * @param type - preselected type to determine list of subTypes
+   * @param onlyActive - if true, filter out types/subtypes where activeFlag !== 'Y'
+   */
+  private mapCaseNoteTypes(caseNoteTypes: CaseNoteType[], type?: string, onlyActive = false) {
+    const typeSubTypeMap: { [key: string]: { value: string; text: string }[] } = {}
+    let subTypes: { value: string; text: string }[] = []
 
-    // Get case notes based on given query params
-    const caseNotesPageData = await this.caseNotesService.get(
-      userToken,
-      prisonerData,
-      queryParams,
-      canDeleteSensitiveCaseNotes,
-    )
-    const showingAll = queryParams.showAll
-
-    // Get staffId to use in conditional logic for amend link
-    const { staffId } = res.locals.user
-
-    // Render page
-    return res.render('pages/caseNotesPage', {
-      pageTitle: 'Case notes',
-      ...mapHeaderData(prisonerData, res.locals.user, 'case-notes'),
-      ...caseNotesPageData,
-      showingAll,
-      hasCaseNotes,
-      addCaseNoteLinkUrl,
-      staffId,
+    const types = caseNoteTypes
+      ?.filter(t => !onlyActive || t.activeFlag === 'Y')
+      .map(t => ({ value: t.code, text: t.description }))
+    caseNoteTypes.forEach(t => {
+      typeSubTypeMap[t.code] = t.subCodes?.map(s => ({ value: s.code, text: s.description }))
```

Comment:
I would recommend a reduce rather than instantiating and then mutating 

`const typeSubTypeMap = caseNoteTypes.reduce...`

Reply:
Yeah, fair point. I normally do like a good reduce, but sometimes it hurts my brain.